### PR TITLE
Scala 2.12 support

### PR DESCRIPTION
**Pull request by logarithm**
This patch adds following:
- Scala 2.12 support and cross build to three scala versions
- SBT, Libraries and plugins update
- Addresses all deprecation warnings
- Removes `loglady` since Logger now is moved to `scala-util` repo
- Replaced `scala_tools.time` by `nscala_time` since `scala_tools.time` doesn't support Scala 2.12 and `nscala_time` is being actively maintained

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 language: scala
 
 scala:
-   - 2.10.5
-   - 2.11.7
+   - 2.10.6
+   - 2.11.8
+   - 2.12.1
```
Comment (drcrallen) [inline .travis.yml]: 🎉

File: build.sbt [hunk 2]
```diff
@@ -57,45 +55,70 @@ parallelExecution in Test := false
 
 testOptions += Tests.Argument(TestFrameworks.JUnit, "-Duser.timezone=UTC")
 
-releaseSettings
+releasePublishArtifactsAction := PgpKeys.publishSigned.value
 
-ReleaseKeys.publishArtifactsAction := PgpKeys.publishSigned.value
+javacOptions ++= Seq("-source", "1.7", "-target", "1.7")
```
Comment (drcrallen) [inline build.sbt]: why is this needed? can you add a comment here?

File: src/main/scala/com/metamx/common/scala/Logger.scala [hunk 3]
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (C) 2012 Daniel Lundin
```
Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/Logger.scala]: Where did this come from?

File: src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala [hunk 4]
```diff
@@ -1,35 +0,0 @@
...
- * limitations under the License.
- */
-
-package com.metamx.common.scala.collection.concurrent
-
-import scala.collection.mutable
-import scala.collection.mutable.ArrayBuffer
-import com.metamx.common.scala.Predef._
-
-object atomic
```
Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala]: Does this really need to be eliminated?

File: .travis.yml [hunk 5]
```diff
@@ -1,7 +1,8 @@
 language: scala
 
 scala:
-   - 2.10.5
-   - 2.11.7
+   - 2.10.6
```
Comment (drcrallen) [inline .travis.yml]: Do we really need to continue to support 2.10?

Comment (logarithm) [inline .travis.yml]: A lot of our services are still on scala 2.10

File: src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala (hunk 4 above)
Comment (logarithm) [inline src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala]: `SynchronizedMap` and `SynchronizedBuffer` both Deprecated in 2.12 and scala developers suggest using `ConcurrentHashMap`, and also we use `SynchronizedBuffer` in serendipity project only, which is not used anymore.

File: build.sbt (hunk 2 above)
Comment (logarithm) [inline build.sbt]: this is added here to make all java sources Java7 compatible. That's a common way to do it in scala projects.

File: src/main/scala/com/metamx/common/scala/Logger.scala (hunk 3 above)
Comment (logarithm) [inline src/main/scala/com/metamx/common/scala/Logger.scala]: This is from https://github.com/metamx/loglady/blob/master/src/main/scala/org/eintr/loglady/Logger.scala

File: src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala (hunk 4 above)
Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala]: since this is OSS removing or breaking the exposed functions needs a major version update. can it be included but `@Deprecated`?

File: src/main/scala/com/metamx/common/scala/Logger.scala (hunk 3 above)
Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/Logger.scala]: we need proper attribution in the `NOTICE` if it came external from MMX

Comment (logarithm) [inline src/main/scala/com/metamx/common/scala/Logger.scala]: Added

File: src/main/scala/com/metamx/common/scala/Logger.scala [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  */
+protected final class LocationAwareLogger(protected val logger: Slf4jLocationAwareLogger) extends Logger {
+  import Slf4jLocationAwareLogger.{ERROR_INT, WARN_INT, INFO_INT, DEBUG_INT, TRACE_INT}
+  private val fqcn = classOf[LocationAwareLogger].getCanonicalName
+  protected val logTrace: LogFunc = logger.log(null, fqcn, TRACE_INT, _, null, _)
+  protected val logDebug: LogFunc = logger.log(null, fqcn, DEBUG_INT, _, null, _)
+  protected val logInfo: LogFunc  = logger.log(null, fqcn, INFO_INT,  _, null, _)
+  protected val logWarn: LogFunc  = logger.log(null, fqcn, WARN_INT,  _, null, _)
+  protected val logError: LogFunc = logger.log(null, fqcn, ERROR_INT, _, null, _)
+}
```
Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/Logger.scala]: newline

File: src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala [hunk 7]
```diff
@@ -32,4 +35,4 @@ object atomic
       }
     }
   }
-}
+}
```
Comment (drcrallen) [inline src/main/scala/com/metamx/common/scala/collection/concurrent/atomic.scala]: newline

**Comment by isimonenko** [comment]
https://issues.scala-lang.org/browse/SI-10244

**Comment by logarithm** [comment]
@isimonenko `javacOptions` are options for Java compiler. So, `javacOptions ++= Seq("-source", "1.7", "-target", "1.7")` means that target version and source version of `java` classes will be 1.7. For scalac it's another parameter: `scalacOptions`, which is set below `javacOptions`.

File: src/main/scala/com/metamx/common/scala/Logging.scala [hunk 8]
```diff
@@ -17,5 +17,26 @@
 package com.metamx.common.scala
 
 trait Logging {
-  @transient lazy val log = Logger(getClass)
+  @transient private var logger: Logger = _
+  @transient @volatile private var initialized = false
```
Comment (isimonenko) [inline src/main/scala/com/metamx/common/scala/Logging.scala]: Is there guarantee that transient field will be initialized  on deserialization?

Comment (isimonenko) [inline src/main/scala/com/metamx/common/scala/Logging.scala]: Actually it will be assigned to a default value which is coincidently also `false`
so it will work as expected but may mislead developers. 
suggest to add comment here
